# КОММЕНТАРИЙ РЕВЬЮВЕРА
Привет Ксения!)

Меня зовут Сибирев Владислав, я осуществлял проверку проекта. Может перейдем сразу на <b>ты</b>? Если это неприемлемо, то обязательно дай знать)

Далее по ходу работы ты сможешь встретить мои комментарии, они будут отмечены разными цветами:

<div class="alert alert-block alert-success">
 ✔️ Похвала и возможности развития
</div>
<div class="alert alert-block alert-warning">
⚠️ Некритичные замечания, но стоит обратить внимание
</div>
<div class="alert alert-block alert-danger">
❌ Грубая ошибка, необходимо исправить
</div>


Заранее прошу не удалять мои комментарии, это важно. Если у тебя возникнут вопросы, обозначь их, я с радостью тебе помогу!)

Как вариант, можешь обозначить их так:
<div class="alert alert-block alert-info">
 Комментарий студента 
            </div>

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b> 

✔️ Так как первая часть является автоматической, то я просто буду в ней давать рекомендации, которые ты можешь использовать в дальнейшем))<br>
P.S. Любой проект нужно начинать с описания)) 
  </div>

## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b> 

✔️ Молодец, что используешь конструкцию __try и except__<br>
P.S. Для следующих проектов нужно импортировать библиотеки в отдельной ячейке) Это позволяет быстро сориентироваться, что нужно для запуска проекта, если кто-то будет запускать его в другом окружении.
  </div>

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b>

✔️ Изучить долю пропущенных пропусков можно еще с помощью `data.isna().mean()*100`
  </div>

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b>

 ✔️  Можно также использовать следующую конструкцию `data['total_income'] = data['total_income'].fillna(data.groupby(['income_type'])['total_income'].transform('median'))`
Функция transform() в принципе полезная https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.transform.html
  </div>

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b>

 ✔️  Все верно, можно еще использовать query(), который позволяет работать с нужными данными) https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.query.html
  </div>

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b>

✔️  Функцию написала правильную, молодец, но применение функции лучше проверять с помощью `value_counts`, вдруг какое то условие не сработало)

  </div>

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [90]:
# почему-то у меня не работает всё что выше, поэтому копирую свой код
import pandas as pd # импортирую библиотеку pandas
data = pd.read_csv('/datasets/data.csv') # читаю таблицу

# цикл для замены пропущенных значений в столбце ['income_type'] на медианное значение по тиу занятости
# на каждом шаге цикла с помощью атрибута loc выбирю строки, в которых в ['income_type'] текущий тип занятости (q) + есть пропуски в ['total_income'] 
for q in data['income_type'].unique():
    data.loc[(data['income_type'] == q) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == q), 'total_income'].median()

# перевожу отрицательные значения в столбце ['days_employed'] методом abs() в положительные
data['days_employed'] = abs(data['days_employed'])   

# удаляю строки, в которых встречаются аномальные значения из датафрейма data.
data = data[(data['children'] != -1) & (data['children'] != 20)] 

# цикл для замены пропущенных значений в столбце ['days_employed'] на медианное значение по тиу занятости ['income_type']
# на каждом шаге цикла с помощью атрибута loc выбирю строки, в которых в ['income_type'] текущий тип занятости (q) + есть пропуски в ['days_employed']
for q in data['income_type'].unique():
    data.loc[(data['income_type'] == q) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == q), 'days_employed'].median()

#привожу в int все значения
data['total_income'] = data['total_income'].astype(int)     
    
#привожу к единому регистру
data['education'] = data['education'].str.lower() # ваш код здесь

#удаляю явные дубликаты
data = data.drop_duplicates()

# создаю функцию по категоризации данных по столбцу ['total_income']
def categorize_income(total_income):
    if total_income <= 30000:
        return 'E'
    if 30001 <= total_income <= 50000:
        return 'D'
    if 50001 <= total_income <= 200000:
        return 'C'
    if 200001 <= total_income <= 1000000:
        return 'B'
    return 'A'

# создаю дополнительный столбец в таблице с категорией зарплат заёмщиков
data['total_income_category'] = data['total_income'].apply(categorize_income) 

# вывожу все уникальные значения целей кредитования в столбце ['purpose']
data['purpose'].unique()

# создаю функцию по распределению целей на категории 
def categorize_purpose(purpose):
    if 'авто' in purpose:
        return 'операции с автомобилем'
    elif 'недвиж' in purpose or 'жил' in purpose:
        return 'операции с недвижимостью'
    elif 'свад' in purpose:
        return 'проведение свадьбы'
    elif 'образов' in purpose:
        return 'получение образования'
    return 'другое'

# создаю дополнительный столбец в таблице с категорией заёмов заёмщиков
data['purpose_category'] = data['purpose'].apply(categorize_purpose) 

data.reset_index(drop = True)

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21326         1    4529.316663         43   среднее             1   
21327         0  343937.404131         67   среднее             1   
21328         1    2113.346888         38   среднее             1   
21329         3    3112.481705         38   среднее             1   
21330         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21326  гражданский брак                 1      F   компаньон     0   
21327   женат / замужем                 0      F   пенсионер     0   
21328  гражданский брак                 1      M   сотрудник     1   
21329   женат / замужем                 0      M   сотрудник     1   
21330   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
0            253875                 покупка жилья                     B   
1            112080       приобретение автомобиля                     C   
2            145885                 покупка жилья                     C   
3            267628    дополнительное образование                     B   
4            158616               сыграть свадьбу                     C   
...             ...                           ...                   ...   
21326        224791             операции с жильем                     B   
21327        155999          сделка с автомобилем                     C   
21328         89672                  недвижимость                     C   
21329        244093  на покупку своего автомобиля                     B   
21330         82047         на покупку автомобиля                     C   

               purpose_category  
0      операции с недвижимостью  
1        операции с автомобилем  
2      операции с недвижимостью  
3         получение образования  
4            проведение свадьбы  
...                         ...  
21326  операции с недвижимостью  
21327    операции с автомобилем  
21328  операции с недвижимостью  
21329    операции с автомобилем  
21330    операции с автомобилем  

[21331 rows x 14 columns]

In [91]:
# создаю функцию для присвоения категории заёмщикам по наличию детей и просрочек

def categorize_debt(children, debt):
    try:
        if children >0 and debt ==1:
            return 'есть дети и есть просрочки'
        elif children >0 and debt ==0:
            return 'есть дети и нет просрочек'
        elif children ==0 and debt ==1:
            return 'нет детей и есть просрочки'
        elif children ==0 and debt ==0:
            return 'нет детей и нет просрочек'
    except:
        return 'нет категории'

In [92]:
# применяю функцию с созданием нового столбца

data['categorize_debt'] = data.apply(lambda data: categorize_debt(data['children'], data['debt']), axis=1)
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   среднее             1   
8         2    6929.865299         35    высшее             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose total_income_category          purpose_category  \
0               покупка жилья                     B  операции с недвижимостью   
1     приобретение автомобиля                     C    операции с автомобилем   
2               покупка жилья                     C  операции с недвижимостью   
3  дополнительное образование                     B     получение образования   
4             сыграть свадьбу                     C        проведение свадьбы   
5               покупка жилья                     B  операции с недвижимостью   
6           операции с жильем                     B  операции с недвижимостью   
7                 образование                     C     получение образования   
8       на проведение свадьбы                     C        проведение свадьбы   
9     покупка жилья для семьи                     C  операции с недвижимостью   

             categorize_debt  
0  есть дети и нет просрочек  
1  есть дети и нет просрочек  
2  нет детей и нет просрочек  
3  есть дети и нет просрочек  
4  нет детей и нет просрочек  
5  нет детей и нет просрочек  
6  нет детей и нет просрочек  
7  нет детей и нет просрочек  
8  есть дети и нет просрочек  
9  нет детей и нет просрочек

In [93]:
#количество всех заемщиков
summary = data.shape[0]

#количество всех заемщиков по категориями
summary_debtors = data.groupby('children')['children'].count()
summary_debtors = summary_debtors.rename('Всего заёмщиков, чел')

#количество заёмщиков с просрочками; делаю через вычитание, чтобы сохранить индексацию строк для построения датафрейма
hadnot_debt = data[data['debt']==0].groupby('children')['children'].count()
had_debt = summary_debtors - hadnot_debt
had_debt = had_debt.rename('Имеют просрочки, чел')

# подсчёт процентного соотношения заемщиков с просрочками к общему количеству всех заёмщиков по категориям
ratio_debtors = had_debt / summary_debtors * 100
ratio_debtors = ratio_debtors.rename('Всего просрочек, %')

# подсчёт процентного соотношения заемщиков с просрочками к общему количеству всех заёмщиков
ratio_debting = had_debt / summary * 100
ratio_debting = ratio_debting.rename('Просрочки среди всех заёмщиков, %')

# процентное соотношение всех просрочников по категориям
ratio_debting_total = ratio_debting / ratio_debtors * 100
ratio_debting_total = ratio_debting_total.rename('Отношение просрочек, %')

In [94]:
# делаю датафрейм новый

index = [ 0, 1, 2, 3, 4, 5] 
columns = [summary_debtors, had_debt, ratio_debtors, ratio_debting, ratio_debting_total]
pivot = pd.DataFrame(index, columns)
pivot

0
Всего заёмщиков, чел Имеют просрочки, чел Всего просрочек, % Просрочки среди всех заёмщиков, % Отношение просрочек, %   
14091                1063                 7.543822           4.983358                          66.058788               0
4808                 444                  9.234609           2.081478                          22.539965               1
2052                 194                  9.454191           0.909474                          9.619802                2
330                  27                   8.181818           0.126576                          1.547044                3
41                   4                    9.756098           0.018752                          0.192209                4
9                    0                    0.000000           0.000000                          NaN                     5

In [136]:
# делаю сводную таблицу
data_pivot = data.pivot_table(index = ['children'], values = 'debt', aggfunc = {'sum', 'count', 'mean'})
data_pivot

count      mean     sum
children                           
0         14091.0  0.075438  1063.0
1          4808.0  0.092346   444.0
2          2052.0  0.094542   194.0
3           330.0  0.081818    27.0
4            41.0  0.097561     4.0
5             9.0  0.000000     0.0


<div class="alert alert-block alert-info">
Мне такой вывод не кажется очень информативным, потому что всё равно надо дополнительно переименовывать столбцы, 
переводить долю кредиторов в проценты, потому что, например 0,09 это какое-то абстрактное, трудноосязаемое заказчику число. 
9% понятнее. Пока не придумала как сделать это проще
            </div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Кому как удобней, в любом случае нужно знать все методы)) 

    
    
  </div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b>

✔️ Отличная работа, в части вывода информации в таблице)) Единственное можно было бы опттмизировать следующим образом: 
    
       data_pivot = data.pivot_table(index = ['children'], values = 'debt', aggfunc = {'sum', 'count', 'mean'})

'mean' это доля кредиторов. <br>

    
    
  </div>

**Вывод**

Самый низкий процент просрочек среди заёмщиков без детей, всего 7,5%. 
Самый высокий процент просрочек среди заёмщиков с одним ребёнком (9,2%) и двумя детьми (9,4%).
Для остальных категорий выборка слишком маленькая, процент просрочек не показателен. 


<div class="alert alert-block alert-warning">
<b>Комментарий ревьюер:</b> 

⚠️ Не бойся использовать в выводах цифры, также нужно сделать акцент на сбалансированности выборок. Примерно выводы будут выглядеть так: 
`люди, которые не имеют детей, лучше всего отдают кредит, доля должников здесь составляет с 7,5%, хуже обстоит дело с кредиторами, кто имеет одного или двух детей, здесь доля невозврата составляет соответственно 9,2% и 9,4%. Остальные категории уже некорректно сравнивать, так как в них очень маленькая выборка`)) <br>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Над выводами поработал хорошо, молодец)) 
    
    
  </div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [145]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.

#количество всех заемщиков
summary = data.shape[0]

#количество всех заемщиков по категориями
summary_debtors = data.groupby('family_status')['family_status'].count()
summary_debtors = summary_debtors.rename('Всего заёмщиков, чел')

#количество заёмщиков с просрочками; делаю через вычитание, чтобы сохранить индексацию строк для построения датафрейма
hadnot_debt = data[data['debt']==0].groupby('family_status')['family_status'].count()
had_debt = summary_debtors - hadnot_debt
had_debt = had_debt.rename('Имеют просрочки, чел')

# подсчёт процентного соотношения заемщиков с просрочками к общему количеству всех заёмщиков по категориям
ratio_debtors = had_debt / summary_debtors * 100
ratio_debtors = ratio_debtors.rename('Всего просрочек, %')

# подсчёт процентного соотношения заемщиков с просрочками к общему количеству всех заёмщиков
ratio_debting = had_debt / summary * 100
ratio_debting = ratio_debting.rename('Просрочки среди всех заёмщиков, %')

# процентное соотношение всех просрочников по категориям
ratio_debting_total = ratio_debting / ratio_debtors * 100
ratio_debting_total = ratio_debting_total.rename('Отношение просрочек, %')


In [146]:
ratio_debtors.sum()

40.3268863752638

In [156]:
# делаю датафрейм новый

index = [ 'Не женат / не замужем', 'в разводе', 'вдовец / вдова', 'гражданский брак', 'женат / замужем'] 
columns = [summary_debtors, had_debt, ratio_debtors, ratio_debting, ratio_debting_total]
pivot = pd.DataFrame(index, columns)
pivot.sort_values(by='Всего заёмщиков, чел', ascending=False)

0
Всего заёмщиков, чел Имеют просрочки, чел Всего просрочек, % Просрочки среди всех заёмщиков, % Отношение просрочек, %                       
12261                927                  7.560558           4.345788                          57.479724                     женат / замужем
4134                 385                  9.313014           1.804885                          19.380245                    гражданский брак
2796                 273                  9.763948           1.279827                          13.107684               Не женат / не замужем
1189                 84                   7.064760           0.393793                          5.574047                            в разводе
951                  63                   6.624606           0.295345                          4.458300                       вдовец / вдова

**Вывод:** 

9% просрочек среди неженатых и живущих гражданским браком, 6% среди вдовствующих граждан. 
Самая платёжеспособная категория заёмщиков состоят в официальном браке. Среди них просрочек всего 7,5%

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюер:</b> 

⚠️ Удели здесь внимание тоже сбалансированности выборок, по большому счету здесь можно сравнивать только три группы))


<div class="alert alert-block alert-info">
Может быть имеет смысл объединить группу вдовствующих и в разводе в категорию "состоял в браке"? В них примерно одинаковое количество людей, можно сказать даже, что с похожим матриальным состоянием (раздел общего имущества или наследство), которые имеют самый низкий процент. При этом по численности будет такой же как не женат/не замужем. 
            </div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Ты как исследователь, можешь изучать любые взаимосвязи, здесь нет никаких препятствий, такое объединение возможно)) 
    
    
  </div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [157]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.


#количество всех заемщиков
summary = data.shape[0]

#количество всех заемщиков по категориями
summary_debtors = data.groupby('total_income_category')['total_income_category'].count()
summary_debtors = summary_debtors.rename('Всего заёмщиков, чел')

#количество заёмщиков с просрочками; делаю через вычитание, чтобы сохранить индексацию строк для построения датафрейма
hadnot_debt = data[data['debt']==0].groupby('total_income_category')['total_income_category'].count()
had_debt = summary_debtors - hadnot_debt
had_debt = had_debt.rename('Имеют просрочки, чел')

# подсчёт процентного соотношения заемщиков с просрочками к общему количеству всех заёмщиков по категориям
ratio_debtors = had_debt / summary_debtors * 100
ratio_debtors = ratio_debtors.rename('Всего просрочек, %')

# подсчёт процентного соотношения заемщиков с просрочками к общему количеству всех заёмщиков
ratio_debting = had_debt / summary * 100
ratio_debting = ratio_debting.rename('Просрочки среди всех заёмщиков, %')

# процентное соотношение всех просрочников по категориям
ratio_debting_total = ratio_debting / ratio_debtors * 100
ratio_debting_total = ratio_debting_total.rename('Отношение просрочек, %')

In [158]:
# делаю датафрейм новый

index = ['A', 'B', 'C', 'D', 'E'] 
columns = [summary_debtors, had_debt, ratio_debtors, ratio_debting, ratio_debting_total]
pivot = pd.DataFrame(index, columns)
pivot.sort_values(by='Всего заёмщиков, чел', ascending=False)

0
Всего заёмщиков, чел Имеют просрочки, чел Всего просрочек, % Просрочки среди всех заёмщиков, % Отношение просрочек, %   
15921                1353                 8.498210           6.342881                          74.637851               C
5014                 354                  7.060231           1.659557                          23.505696               B
349                  21                   6.017192           0.098448                          1.636116                D
25                   2                    8.000000           0.009376                          0.117200                A
22                   2                    9.090909           0.009376                          0.103136                E

**Вывод:** 

8,5% просрочек среди заёмщиков с доходом от 50 до 200 тысяч. При этом, это 75% всех займов. 

7% просрочек среди заёмщиков с доходом от 200 до миллиона. При этом, это только 23% всех азаймов.
 
Остальные категории имеют слишком маленькую выборку. 


<div class="alert alert-block alert-warning">
<b>Комментарий ревьюер:</b> 

⚠️ Здесь тоже можно сравнить только две группы B и С <br>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Две самые большие группы сравнила, молодец)) 
    
    
  </div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [108]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.

#количество всех заемщиков
summary = data.shape[0]

#количество всех заемщиков по категориями
summary_debtors = data.groupby('purpose_category')['purpose_category'].count()
summary_debtors = summary_debtors.rename('Всего заёмщиков, чел')

#количество заёмщиков с просрочками; делаю через вычитание, чтобы сохранить индексацию строк для построения датафрейма
hadnot_debt = data[data['debt']==0].groupby('purpose_category')['purpose_category'].count()
had_debt = summary_debtors - hadnot_debt
had_debt = had_debt.rename('Имеют просрочки, чел')

# подсчёт процентного соотношения заемщиков с просрочками к общему количеству всех заёмщиков по категориям
ratio_debtors = had_debt / summary_debtors * 100
ratio_debtors = ratio_debtors.rename('Всего просрочек, %')

# подсчёт процентного соотношения заемщиков с просрочками к общему количеству всех заёмщиков
ratio_debting = had_debt / summary * 100
ratio_debting = ratio_debting.rename('Просрочки среди всех заёмщиков, %')

# процентное соотношение всех просрочников по категориям
ratio_debting_total = ratio_debting / ratio_debtors * 100
ratio_debting_total = ratio_debting_total.rename('Отношение просрочек, %')

In [114]:
# Количество заёмщиков на операции с недвижимостью по виду деятельности

debtors_income_type = data.loc[(data['purpose_category']=='операции с недвижимостью')].groupby('income_type')['income_type'].count()
summary_debtors_income_type = debtors_income_type.sum()
ratio = debtors_income_type / summary_debtors_income_type * 100
ratio


income_type
безработный         0.018603
госслужащий         7.004000
компаньон          23.541996
пенсионер          17.496047
предприниматель     0.009301
сотрудник          51.920752
студент             0.009301
Name: income_type, dtype: float64

In [111]:
# Количество заёмщиков на операции с недвижимостью по категории дохода

debtors_income = data.loc[(data['purpose_category']=='операции с недвижимостью')].groupby('total_income_category')['total_income_category'].count()
summary_debtors_income = debtors_income.sum()
ratio = debtors_income / summary_debtors_income * 100
ratio

total_income_category
A     0.158125
B    23.793136
C    74.383778
D     1.562645
E     0.102316
Name: total_income_category, dtype: float64

In [109]:
# делаю датафрейм новый

index = [ 'операции с автомобилем', 'операции с недвижимостью', 'получение образования', 'проведение свадьбы'] 
columns = [summary_debtors, had_debt, ratio_debtors, ratio_debting, ratio_debting_total]
pivot = pd.DataFrame(index, columns)
pivot

0
Всего заёмщиков, чел Имеют просрочки, чел Всего просрочек, % Просрочки среди всех заёмщиков, % Отношение просрочек, %                          
4279                 400                  9.347978           1.875205                          20.060007                 операции с автомобилем
10751                780                  7.255139           3.656650                          50.400825               операции с недвижимостью
3988                 369                  9.252758           1.729877                          18.695795                  получение образования
2313                 183                  7.911803           0.857906                          10.843373                     проведение свадьбы

**Вывод:** 


50% ипотечных займов. Как правило ипотеку берут заёмщики с уверенным уровнем дохода и стабильной зарплатой, так как это долгосрочный крупный займ и требует стабильного финансового положения. Это подтверждают цифры -  98% всех заёмщиков на ипотеку имеют уровень дохода от 50 000 рублей, из них 75% официально трудоустроены в компанях или являются совладельцами бизнеса, просрочек в этой категории меньше всего - 7,25%.

По 20 и 18% в категориях операции с автомобилями и получение образования соответственно имеют просрочки по 9%. 

Меньше всего берут займ на свадьбу, всего 10% и уровень просрочек примерно 8%. 




<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера:</b> 

⚠️ Почему по твоему самые надежные кредиторы те, кто берет ипотеку?


<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Все верно, к тому же к ним предъявляются самые жесткие требования)) 
    
    
  </div>

*****************

Мой вопрос 1: кто чаще берёт кредиты: мужчины или женщины?

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера:</b> 

❌ Здесь у тебя код ломается((

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Молодец, что проанализировала данные, не входящие в задание))
    
    
  </div>

In [36]:
answer1 = data.groupby('gender')['gender'].count()
print(answer1)
women = data.loc[data['gender'] =='F'].groupby('gender')['gender'].count()
men = data.loc[data['gender'] =='M'].groupby('gender')['gender'].count()
ratio = int(women) / int(men)
ratio

gender
F      14092
M       7238
XNA        1
Name: gender, dtype: int64


1.9469466703509257

Ответ: кредиты в два раза чаще берут женщины. 

Мой вопрос 2: медианные значения у мужчин и у женщин?

In [17]:
med = data.groupby('gender').median()
med

children  days_employed  dob_years  education_id  family_status_id  \
gender                                                                       
F            0.0    2385.547863       44.0           1.0               0.0   
M            0.0    1572.328285       40.0           1.0               0.0   
XNA          0.0    2358.600502       24.0           2.0               1.0   

        debt  total_income  
gender                      
F        0.0      138933.5  
M        0.0      163802.5  
XNA      0.0      203905.0

Ответ: 

Зарплата Ж = 138 тысяч, М = 163 тысячи,

Возраст Ж = 40 лет, М = 44, 

Обе категории имеют высшее образование, не имеют детей и не состоят в зарегистрированных отношениях. 

Третий пол - возможно техническая ошибка. 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

Пропуски в исходных данных могут быть связаны с технической ошибкой, например сбой оборудования или ошибка в программе обработки данных, могут быть связаны с человеческим фактором, например кто-то забыл внести значение 
и психологическим фактором, кто-то не готов указывать о себе какие-то данные. 



<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b>

✔️ Здесь подробно все описал, молодец)) 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

Медиана имеет срединное значение из списка, а не среднее значение чисел в списке, поэтому она может более точно показать наиболее распространённое значение в списке, что более правдоподобно для массива данных. 

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b>

✔️ Все правильно медиана не подтвержена влиянию выбросов)) 

### Шаг 4: общий вывод.

#### Есть ли зависимость между количеством детей и возвратом кредита в срок?

Ответ: количество детей в целом не влияет на платёжеспособность заёмщиков, средний процент просрочек на уровне 9%, однако наименее ответственны граждане без детей, они чаще берут кредиты и больше имеют просрочек.


#### Есть ли зависимость между семейным положением и возвратом кредита в срок?

Ответ: семейные люди имеют платёжеспособность выше, даже если брак уже не действителен или не был зарегистрирован, чем люди не состоящие в отношениях. 


#### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Ответ: заёмщики с самым большим доходом не гарантируют, что всё будет заплачено во время. 



#### Как разные цели кредита влияют на его возврат в срок?

Ответ:

На образование и покупку автомобиля берут 38% заёмщиков, при этом 9% имеют просрочки. 
50% займов берут на операции с недвижимостью, при этом самый низкий процет просрочки 7,5%. 

50% ипотечных займов. Как правило ипотеку берут заёмщики с уверенным уровнем дохода и стабильной зарплатой, так как это долгосрочный крупный займ и требует стабильного финансового положения. Это подтверждают цифры -  98% всех заёмщиков на ипотеку имеют уровень дохода от 50 000 рублей, из них 75% официально трудоустроены в компанях или являются совладельцами бизнеса, просрочек в этой категории меньше всего - 7,25%.

По 20 и 18% в категориях операции с автомобилями и получение образования соответственно имеют просрочки по 9%. 

Меньше всего берут займ на свадьбу, всего 10% и уровень просрочек примерно 8%. 



Даешь рекомендацией банку по улучшению сбора данных и скоринга.






Женщины берут займы в два раза больше, чем мужчины, при этом зарплата у женщин на 18% меньше. 


Можно сделать вывод, что примерно 9% граждан имеют просрочки независимо от уровня дохода, цели кредита или семейного положения. 


<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Выводы переработала хорошо, молодец)) 
    
    
  </div>

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера:</b> 

❌ Выводы делаем детально, что должно быть в них:
- Отвечаешь на вопрос "Есть ли зависимость между количеством детей и возвратом кредита в срок?"
- Отвечаешь на вопрос  "Есть ли зависимость между семейным положением и возвратом кредита в срок?"
- Отвечаешь на вопрос  "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?"
- Отвечаешь на вопрос  "Как разные цели кредита влияют на его возврат в срок?"
- Даешь рекомендацией банку по улучшению сбора данных и скоринга.

Общий вывод - резюме этой работы. Большинство заказчиков посмотрят сначала его (а некоторые только его). Чтобы выводы не огорошивали заказчика своей очевидностью, стоит добавить в них цифр, всяких корреляций, процентов итд :) Сделай так, чтобы вся найденная в данных информация была отражена в выводе 

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b>
    
Ксения, отлично поработала, видно что умеешь работать с кодом, надо будет избавиться от мультииндексов в таблицах и  поработать над выводами, в остальном все отлично)) 
<br> Жду исправленного варианта)
  </div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Ксения, отличная работа, все исправила, можно двигаться дальше, успехов в дальнейшем обучении)) 
    
    
  </div>